server/schedule/merge_checker.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedule
+
+import (
+	"github.com/pingcap/pd/server/core"
+	"github.com/pingcap/pd/server/namespace"
+	log "github.com/sirupsen/logrus"
+)
+
+// MergeChecker ensures region to merge with adjacent region when size is small
+type MergeChecker struct {
+	cluster    Cluster
+	classifier namespace.Classifier
+}
+
+// NewMergeChecker creates a merge checker.
+func NewMergeChecker(cluster Cluster, classifier namespace.Classifier) *MergeChecker {
+	return &MergeChecker{
+		cluster:    cluster,
+		classifier: classifier,
+	}
+}
+
+// Check verifies a region's replicas, creating an Operator if need.
+func (m *MergeChecker) Check(region *core.RegionInfo) (*Operator, *Operator) {
+	checkerCounter.WithLabelValues("merge_checker", "check").Inc()
+
+	// when pd just started, it will load region meta from etcd
+	// but the size for these loaded region info is 0
+	// pd don't the real size of one region until the first heartbeat of the region
+	// thus here when size is 0, just skip.
+	if region.ApproximateSize == 0 {
+		checkerCounter.WithLabelValues("merge_checker", "skip").Inc()
+		return nil, nil
+	}
+
+	// region size is not small enough
+	if region.ApproximateSize > int64(m.cluster.GetMaxMergeRegionSize()) {
+		checkerCounter.WithLabelValues("merge_checker", "no_need").Inc()
+		return nil, nil
+	}
+
+	// skip region has down peers
+	if len(region.DownPeers) > 0 {
+		checkerCounter.WithLabelValues("merge_checker", "down_peers").Inc()
+		return nil, nil
+	}
+
+	// skip region has pending peers
+	if len(region.PendingPeers) > 0 {
+		checkerCounter.WithLabelValues("merge_checker", "pending_peers").Inc()
+		return nil, nil
+	}
+
+	// skip hot region
+	if m.cluster.IsRegionHot(region.GetId()) {
+		checkerCounter.WithLabelValues("merge_checker", "hot_region").Inc()
+		return nil, nil
+	}
```

Comment:
I think we can also skip the region if its replica count is different from default configuration.